Dashboard Test Sometimes Hangs

I see this specific test in `test_dashboard` hang quite often for
https://github.com/BLAST-ImpactX/impactx/blob/25.08/tests/python/dashboard/test_dashboard.py#L76-L89

Failing with
```
TimeoutError: state['total_length'] never became '3.0m' after 60 seconds (last value: '2.75m').
```
[job-logs.txt](https://github.com/user-attachments/files/21902118/job-logs.txt)

I can usually solve it with a test restart in CI, but that is of course cumbersome and does not work in unsupervised situations, eg.., when we release software in a package manager like Conda/Spack.

This seems to happen more often for the `GCC w/ SIMD w/o MPI w/ Python` test.

I have the following ideas how we could improve this test:

- Since this is a floating point comparison, should we introduce a small epsilon instead of expecting `3.0` exactly? (usually that is the best practice, and we will need this for single precision, too.)
- Does the test need a bit more time to load the GUI up, e.g., before it starts to set the lattice elements? A `0.25m` gap missing [could be the `ds1` or the `ds5` element](https://github.com/BLAST-ImpactX/impactx/blob/25.08/tests/python/dashboard/test_dashboard.py#L48-L66), maybe we can print the lattice for this value to find out what is missing? Do we need selenium to wait for the GUI to update after each `dashboard.add_lattice_element` or `dashboard.set_input(...)`?

Related: https://github.com/BLAST-ImpactX/impactx/pull/1094#issuecomment-3194799179
```
>       raise TimeoutError(
            f"state['{state_name}'] never became '{expected_input}' after {timeout} seconds "
            f"(last value: '{value}').\n"
            f"curr_view_details_log: {self.get_state('curr_view_details_log')}"
        )
E       TimeoutError: state['total_steps'] never became '125' after 60 seconds (last value: '101').
```

Got another one in #1111 ([run](https://github.com/BLAST-ImpactX/impactx/actions/runs/17220549134/job/48854689921?pr=1111)): [job-logs.txt](https://github.com/user-attachments/files/21977477/job-logs.txt)
```
 >       raise TimeoutError(
             f"state['{state_name}'] never became '{expected_input}' after {timeout} seconds "
             f"(last value: '{value}').\n"
             f"curr_view_details_log: {self.get_state('curr_view_details_log')}"
         )
 E       TimeoutError: state['total_length'] never became '3.0m' after 60 seconds (last value: '2.75m').
 E       curr_view_details_log:
 
 /home/runner/work/impactx/impactx/tests/python/dashboard/utils.py:226: TimeoutError
 =========================== short test summary info ============================
 FAILED ../../../tests/python/dashboard/test_dashboard.py::test_dashboard - TimeoutError: state['total_length'] never became '3.0m' after 60 seconds (last value: '2.75m').
 curr_view_details_log:
 =================== 1 failed, 2 passed in 102.67s (0:01:42) ====================
```

--- a/.github/workflows/macos.yml
+++ b/.github/workflows/macos.yml
@@ -85,7 +85,8 @@ jobs:
       with:
         name: dashboard-screenshots-macos
         path: |
-          build/bin/pytest.ImpactX.dashboard/screenshots/**/*.png
+          build/**/pytest.ImpactX.dashboard/screenshots/**/*.png
+          build/**/screenshots/**/*.png
         if-no-files-found: ignore
 
     - name: run installed python module

--- a/.github/workflows/macos.yml
+++ b/.github/workflows/macos.yml
@@ -79,6 +79,15 @@ jobs:
       run: |
         ctest --test-dir build --output-on-failure -E pytest.AMReX
 
+    - name: Upload dashboard screenshots
+      if: always()
+      uses: actions/upload-artifact@v4
+      with:
+        name: dashboard-screenshots-macos
+        path: |
+          build/bin/pytest.ImpactX.dashboard/screenshots/**/*.png
+        if-no-files-found: ignore
+
     - name: run installed python module
       run: |
         cmake --build build --target pip_install

--- a/.github/workflows/ubuntu.yml
+++ b/.github/workflows/ubuntu.yml
@@ -75,7 +75,8 @@ jobs:
       with:
         name: dashboard-screenshots-ubuntu
         path: |
-          build/bin/pytest.ImpactX.dashboard/screenshots/**/*.png
+          build/**/pytest.ImpactX.dashboard/screenshots/**/*.png
+          build/**/screenshots/**/*.png
         if-no-files-found: ignore
 
     - name: run installed app

--- a/.github/workflows/ubuntu.yml
+++ b/.github/workflows/ubuntu.yml
@@ -69,6 +69,15 @@ jobs:
       run: |
         ctest --test-dir build --output-on-failure --label-exclude slow
 
+    - name: Upload dashboard screenshots
+      if: always()
+      uses: actions/upload-artifact@v4
+      with:
+        name: dashboard-screenshots-ubuntu
+        path: |
+          build/bin/pytest.ImpactX.dashboard/screenshots/**/*.png
+        if-no-files-found: ignore
+
     - name: run installed app
       run: |
         cmake --build build --target install

--- a/tests/python/CMakeLists.txt
+++ b/tests/python/CMakeLists.txt
@@ -32,6 +32,9 @@ set_property(TEST ${dashboard_test_name} PROPERTY LABELS "dashboard")
 set_property(TEST ${pytest_name} APPEND PROPERTY ENVIRONMENT "OMP_NUM_THREADS=2")
 set_property(TEST ${pytest_name}.dashboard APPEND PROPERTY ENVIRONMENT "OMP_NUM_THREADS=2")
 
+# Save dashboard test screenshots to a known path for CI artifact collection
+set_property(TEST ${pytest_name}.dashboard APPEND PROPERTY ENVIRONMENT "IMPACTX_SCREENSHOT_DIR=${pytest_rundir}.dashboard/screenshots")
+
 
 #   set PYTHONPATH and PATH (for .dll files)
 impactx_test_set_pythonpath(${pytest_name})

--- a/tests/python/dashboard/conftest.py
+++ b/tests/python/dashboard/conftest.py
@@ -1,5 +1,3 @@
-import os
-import time
 import pytest
 from seleniumbase import SB
 
@@ -8,18 +6,10 @@ from .utils import (
     start_dashboard,
     wait_for_interaction_ready,
     wait_for_server_ready,
+    save_failure_screenshot,
 )
 
 
-@pytest.hookimpl(hookwrapper=True, tryfirst=True)
-def pytest_runtest_makereport(item, call):
-    """Attach test phase reports to the node for later inspection.
-
-    Allows fixtures to check failures in teardown via request.node.rep_call.
-    """
-    outcome = yield
-    rep = outcome.get_result()
-    setattr(item, "rep_" + rep.when, rep)
 
 
 @pytest.fixture(scope="session")
@@ -55,15 +45,5 @@ def reset_dashboard_inputs(dashboard, request):
     dashboard.sb.click("#reset_all_inputs_button")
     # Teardown: on failure, save a screenshot for debugging
     yield
-    rep = getattr(request.node, "rep_call", None)
-    if rep and rep.failed:
-        try:
-            os.makedirs("screenshots", exist_ok=True)
-            ts = time.strftime("%Y%m%d-%H%M%S")
-            name = f"{request.node.name}_{ts}.png"
-            path = os.path.abspath(os.path.join("screenshots", name))
-            # Use the underlying WebDriver to capture the screen
-            dashboard.sb.driver.save_screenshot(path)
-            print(f"Saved failed test screenshot: {path}")
-        except Exception:
-            pass
+    # Always save a screenshot after each test for now (CI artifact)
+    save_failure_screenshot(dashboard, request)

--- a/tests/python/dashboard/conftest.py
+++ b/tests/python/dashboard/conftest.py
@@ -3,13 +3,11 @@ from seleniumbase import SB
 
 from .utils import (
     DashboardTester,
+    save_failure_screenshot,
     start_dashboard,
     wait_for_interaction_ready,
     wait_for_server_ready,
-    save_failure_screenshot,
 )
-
-
 
 
 @pytest.fixture(scope="session")
@@ -38,12 +36,13 @@ def dashboard():
 def reset_dashboard_inputs(dashboard, request):
     """
     Resets the dashboard to its default state before each test.
-
     This ensures all tests start from a clean, consistent baseline.
     """
+ 
     dashboard.sb.click("#Input_route")
     dashboard.sb.click("#reset_all_inputs_button")
     # Teardown: on failure, save a screenshot for debugging
     yield
+
     # Always save a screenshot after each test for now (CI artifact)
     save_failure_screenshot(dashboard, request)

--- a/tests/python/dashboard/conftest.py
+++ b/tests/python/dashboard/conftest.py
@@ -38,7 +38,7 @@ def reset_dashboard_inputs(dashboard, request):
     Resets the dashboard to its default state before each test.
     This ensures all tests start from a clean, consistent baseline.
     """
- 
+
     dashboard.sb.click("#Input_route")
     dashboard.sb.click("#reset_all_inputs_button")
     # Teardown: on failure, save a screenshot for debugging

--- a/tests/python/dashboard/conftest.py
+++ b/tests/python/dashboard/conftest.py
@@ -1,3 +1,5 @@
+import os
+import time
 import pytest
 from seleniumbase import SB
 
@@ -7,6 +9,17 @@ from .utils import (
     wait_for_interaction_ready,
     wait_for_server_ready,
 )
+
+
+@pytest.hookimpl(hookwrapper=True, tryfirst=True)
+def pytest_runtest_makereport(item, call):
+    """Attach test phase reports to the node for later inspection.
+
+    Allows fixtures to check failures in teardown via request.node.rep_call.
+    """
+    outcome = yield
+    rep = outcome.get_result()
+    setattr(item, "rep_" + rep.when, rep)
 
 
 @pytest.fixture(scope="session")
@@ -32,7 +45,7 @@ def dashboard():
 
 
 @pytest.fixture(autouse=True)
-def reset_dashboard_inputs(dashboard):
+def reset_dashboard_inputs(dashboard, request):
     """
     Resets the dashboard to its default state before each test.
 
@@ -40,3 +53,17 @@ def reset_dashboard_inputs(dashboard):
     """
     dashboard.sb.click("#Input_route")
     dashboard.sb.click("#reset_all_inputs_button")
+    # Teardown: on failure, save a screenshot for debugging
+    yield
+    rep = getattr(request.node, "rep_call", None)
+    if rep and rep.failed:
+        try:
+            os.makedirs("screenshots", exist_ok=True)
+            ts = time.strftime("%Y%m%d-%H%M%S")
+            name = f"{request.node.name}_{ts}.png"
+            path = os.path.abspath(os.path.join("screenshots", name))
+            # Use the underlying WebDriver to capture the screen
+            dashboard.sb.driver.save_screenshot(path)
+            print(f"Saved failed test screenshot: {path}")
+        except Exception:
+            pass

--- a/tests/python/dashboard/test_dashboard.py
+++ b/tests/python/dashboard/test_dashboard.py
@@ -6,6 +6,7 @@ Authors: Parthib Roy
 License: BSD-3-Clause-LBNL
 """
 
+
 def test_dashboard(dashboard):
     """
     End-to-end test of the ImpactX dashboard by directly setting inputs via UI,

--- a/tests/python/dashboard/test_dashboard.py
+++ b/tests/python/dashboard/test_dashboard.py
@@ -45,6 +45,9 @@ def test_dashboard(dashboard):
     for name in ["Drift", "Quad", "Drift", "Quad", "Drift"]:
         dashboard.add_lattice_element(name)
 
+    # Ensure the lattice list is fully populated before setting parameters
+    dashboard.assert_state("total_elements", 5)
+
     LATTICE_PARAMS = {
         "ds1": 0.25,
         "nslice1": "ns",

--- a/tests/python/dashboard/test_dashboard.py
+++ b/tests/python/dashboard/test_dashboard.py
@@ -1,3 +1,11 @@
+"""
+This file is part of ImpactX
+
+Copyright 2025 ImpactX contributors
+Authors: Parthib Roy
+License: BSD-3-Clause-LBNL
+"""
+
 def test_dashboard(dashboard):
     """
     End-to-end test of the ImpactX dashboard by directly setting inputs via UI,

--- a/tests/python/dashboard/test_lattice_variable_handler.py
+++ b/tests/python/dashboard/test_lattice_variable_handler.py
@@ -111,6 +111,7 @@ def test_lattice_variable_handler(dashboard):
     dashboard.sb.click("#lattice_settings")
     for i in range(len(ALL_VARIABLES) - 1):
         dashboard.sb.click(f"#add_variable_button_{i + 1}")
+        dashboard.sb.assert_element(f"#variable_name_{i + 2}")
 
     # Enter variables on the dashboard
     for i, var in enumerate(ALL_VARIABLES, start=1):

--- a/tests/python/dashboard/test_lattice_variable_handler.py
+++ b/tests/python/dashboard/test_lattice_variable_handler.py
@@ -55,8 +55,17 @@ def assert_lattice_param_sim_input(
     # Simple retry logic similar to assert_state
     for i in range(10):
         current_value = get_sim_input()
-        if current_value == expected_value:
-            return
+        try:
+            if isinstance(expected_value, (int, float)) and isinstance(
+                current_value, (int, float)
+            ):
+                if current_value == pytest.approx(expected_value, rel=1e-12, abs=1e-12):
+                    return
+            elif current_value == expected_value:
+                return
+        except Exception:
+            if current_value == expected_value:
+                return
         time.sleep(1)
 
     raise AssertionError(

--- a/tests/python/dashboard/test_lattice_variable_handler.py
+++ b/tests/python/dashboard/test_lattice_variable_handler.py
@@ -9,7 +9,9 @@ License: BSD-3-Clause-LBNL
 import time
 
 import pytest
+
 from .utils import APPROX_TOL, TIMEOUT
+
 
 def lattice_value(state, index: int, param_name: str) -> float:
     """

--- a/tests/python/dashboard/test_lattice_variable_handler.py
+++ b/tests/python/dashboard/test_lattice_variable_handler.py
@@ -9,7 +9,7 @@ License: BSD-3-Clause-LBNL
 import time
 
 import pytest
-from .utils import TIMEOUT
+from .utils import APPROX_TOL, TIMEOUT
 
 def lattice_value(state, index: int, param_name: str) -> float:
     """
@@ -59,7 +59,7 @@ def assert_lattice_param_sim_input(
             if isinstance(expected_value, (int, float)) and isinstance(
                 current_value, (int, float)
             ):
-                if current_value == pytest.approx(expected_value, rel=1e-12, abs=1e-12):
+                if current_value == pytest.approx(expected_value, **APPROX_TOL):
                     return
             elif current_value == expected_value:
                 return

--- a/tests/python/dashboard/test_lattice_variable_handler.py
+++ b/tests/python/dashboard/test_lattice_variable_handler.py
@@ -9,7 +9,7 @@ License: BSD-3-Clause-LBNL
 import time
 
 import pytest
-
+from .utils import TIMEOUT
 
 def lattice_value(state, index: int, param_name: str) -> float:
     """
@@ -53,7 +53,7 @@ def assert_lattice_param_sim_input(
         return None
 
     # Simple retry logic similar to assert_state
-    for i in range(10):
+    for i in range(TIMEOUT):
         current_value = get_sim_input()
         try:
             if isinstance(expected_value, (int, float)) and isinstance(
@@ -111,7 +111,6 @@ def test_lattice_variable_handler(dashboard):
     dashboard.sb.click("#lattice_settings")
     for i in range(len(ALL_VARIABLES) - 1):
         dashboard.sb.click(f"#add_variable_button_{i + 1}")
-        dashboard.sb.assert_element(f"#variable_name_{i + 2}")
 
     # Enter variables on the dashboard
     for i, var in enumerate(ALL_VARIABLES, start=1):

--- a/tests/python/dashboard/test_python_import.py
+++ b/tests/python/dashboard/test_python_import.py
@@ -1,3 +1,12 @@
+"""
+This file is part of ImpactX
+
+Copyright 2025 ImpactX contributors
+Authors: Parthib Roy
+License: BSD-3-Clause-LBNL
+"""
+import pytest
+
 def test_python_import(dashboard):
     """
     End-to-end test of the ImpactX dashboard by importing input values from a Python file.
@@ -60,6 +69,6 @@ def test_python_import(dashboard):
     # Check input values
     for element_id, expected_value in DISTRIBUTION_VALUES + LATTICE_CONFIGURATION:
         actual_value = float(dashboard.sb.get_value(element_id))
-        assert actual_value == expected_value, (
-            f"{element_id}: expected {expected_value}, got {actual_value}"
-        )
+        assert actual_value == pytest.approx(
+            expected_value, rel=1e-12, abs=1e-12
+        ), f"{element_id}: expected {expected_value}, got {actual_value}"

--- a/tests/python/dashboard/test_python_import.py
+++ b/tests/python/dashboard/test_python_import.py
@@ -6,6 +6,7 @@ Authors: Parthib Roy
 License: BSD-3-Clause-LBNL
 """
 import pytest
+from .utils import APPROX_TOL
 
 def test_python_import(dashboard):
     """
@@ -70,5 +71,5 @@ def test_python_import(dashboard):
     for element_id, expected_value in DISTRIBUTION_VALUES + LATTICE_CONFIGURATION:
         actual_value = float(dashboard.sb.get_value(element_id))
         assert actual_value == pytest.approx(
-            expected_value, rel=1e-12, abs=1e-12
+            expected_value, **APPROX_TOL
         ), f"{element_id}: expected {expected_value}, got {actual_value}"

--- a/tests/python/dashboard/test_python_import.py
+++ b/tests/python/dashboard/test_python_import.py
@@ -5,8 +5,11 @@ Copyright 2025 ImpactX contributors
 Authors: Parthib Roy
 License: BSD-3-Clause-LBNL
 """
+
 import pytest
+
 from .utils import APPROX_TOL
+
 
 def test_python_import(dashboard):
     """
@@ -70,6 +73,6 @@ def test_python_import(dashboard):
     # Check input values
     for element_id, expected_value in DISTRIBUTION_VALUES + LATTICE_CONFIGURATION:
         actual_value = float(dashboard.sb.get_value(element_id))
-        assert actual_value == pytest.approx(
-            expected_value, **APPROX_TOL
-        ), f"{element_id}: expected {expected_value}, got {actual_value}"
+        assert actual_value == pytest.approx(expected_value, **APPROX_TOL), (
+            f"{element_id}: expected {expected_value}, got {actual_value}"
+        )

--- a/tests/python/dashboard/utils.py
+++ b/tests/python/dashboard/utils.py
@@ -160,7 +160,7 @@ class DashboardTester:
         Purpose:
         This makes the page behave exactly as if a real user typed into the field
         and then clicked out, ensuring the app updates correctly.
-    """
+        """
         js = (
             "var el = document.getElementById(arguments[0]);"
             "if (!el) { throw new Error('element not found'); }"

--- a/tests/python/dashboard/utils.py
+++ b/tests/python/dashboard/utils.py
@@ -14,6 +14,7 @@ from pathlib import Path
 
 from selenium.common.exceptions import TimeoutException
 import os
+import pytest
 import time as _time
 
 TIMEOUT = 120
@@ -215,8 +216,15 @@ class DashboardTester:
                 value = None
 
             if isinstance(expected_input, (int, float)):
-                if value is not None and float(value) == float(expected_input):
-                    return
+                try:
+
+                    v_num = None if value is None else float(value)
+                    if v_num is not None and v_num == pytest.approx(
+                        float(expected_input), rel=1e-12, abs=1e-12
+                    ):
+                        return
+                except (TypeError, ValueError):
+                    pass
             elif value == expected_input:
                 return
 

--- a/tests/python/dashboard/utils.py
+++ b/tests/python/dashboard/utils.py
@@ -12,8 +12,8 @@ import sys
 import time
 from pathlib import Path
 
-from selenium.common.exceptions import TimeoutException
 import pytest
+from selenium.common.exceptions import TimeoutException
 
 TIMEOUT = 120
 APPROX_TOL = {"rel": 1e-12, "abs": 1e-12}
@@ -217,7 +217,6 @@ class DashboardTester:
 
             if isinstance(expected_input, (int, float)):
                 try:
-
                     v_num = None if value is None else float(value)
                     if v_num is not None and v_num == pytest.approx(
                         float(expected_input), **APPROX_TOL
@@ -252,7 +251,9 @@ class DashboardTester:
         return self.sb.execute_script(js_script, state_name)
 
 
-def save_failure_screenshot(dashboard, request, directory: str | None = None) -> str | None:
+def save_failure_screenshot(
+    dashboard, request, directory: str | None = None
+) -> str | None:
     """
     Save a screenshot PNG for the current test and return the absolute path.
 

--- a/tests/python/dashboard/utils.py
+++ b/tests/python/dashboard/utils.py
@@ -139,16 +139,41 @@ class DashboardTester:
         :param new_input: New value to set for the input element.
         """
         try:
-            self.sb.execute_script(
-                f'document.getElementById("{element_id}").value = "{new_input}";'
-            )
-            self.sb.execute_script(
-                f'document.getElementById("{element_id}").dispatchEvent(new Event("input"));'
-            )
+            self._commit_input_js(element_id, new_input)
         except Exception as error:
             raise Exception(
                 f"Unable to set input for lattice element '{element_id}': {str(error)}"
             )
+
+    def _commit_input_js(self, element_id: str, new_input) -> None:
+        """
+        Update a Vuetify text input field in the browser using JavaScript.
+
+        What happens step by step:
+        1. Find the input element by its ID.
+        2. Focus the element (like clicking inside it so it’s active).
+        3. Set the element’s value to the new text.
+        4. Trigger an "input" event so the web app reacts as if you typed the text.
+        5. Trigger a "change" event so the app thinks you finished editing.
+        6. Blur the element (like clicking away so it’s no longer active).
+
+        Purpose:
+        This makes the page behave exactly as if a real user typed into the field
+        and then clicked out, ensuring the app updates correctly.
+    """
+        js = (
+            "var el = document.getElementById(arguments[0]);"
+            "if (!el) { throw new Error('element not found'); }"
+            "el.focus();"
+            "el.value = arguments[1];"
+            "el.dispatchEvent(new InputEvent('input', {bubbles:true,cancelable:true}));"
+            "el.dispatchEvent(new Event('change', {bubbles:true}));"
+            "el.blur();"
+            "return true;"
+        )
+        successful_input = self.sb.execute_script(js, element_id, str(new_input))
+        if not successful_input:
+            raise RuntimeError("failed to commit value via JS")
 
     def set_state(self, state_name: str, state_value):
         """

--- a/tests/python/dashboard/utils.py
+++ b/tests/python/dashboard/utils.py
@@ -13,9 +13,7 @@ import time
 from pathlib import Path
 
 from selenium.common.exceptions import TimeoutException
-import os
 import pytest
-import time as _time
 
 TIMEOUT = 120
 APPROX_TOL = {"rel": 1e-12, "abs": 1e-12}
@@ -256,7 +254,7 @@ class DashboardTester:
 
 def save_failure_screenshot(dashboard, request, directory: str | None = None) -> str | None:
     """
-    Save a screenshot PNG for the current test and print the absolute path.
+    Save a screenshot PNG for the current test and return the absolute path.
 
     - Respects env var `IMPACTX_SCREENSHOT_DIR` to override the output folder.
     - Falls back to `directory` arg or `screenshots/` in the current CWD.
@@ -265,12 +263,9 @@ def save_failure_screenshot(dashboard, request, directory: str | None = None) ->
     base_dir = os.environ.get("IMPACTX_SCREENSHOT_DIR", directory or "screenshots")
     try:
         os.makedirs(base_dir, exist_ok=True)
-        ts = _time.strftime("%Y%m%d-%H%M%S")
-        # Keep name filesystem-friendly
-        name = f"{request.node.name}_{ts}.png".replace(os.sep, "_")
+        name = f"{request.node.name}.png".replace(os.sep, "_")
         path = os.path.abspath(os.path.join(base_dir, name))
         dashboard.sb.driver.save_screenshot(path)
-        print(f"Saved test screenshot: {path}")
         return path
     except Exception:
         return None

--- a/tests/python/dashboard/utils.py
+++ b/tests/python/dashboard/utils.py
@@ -14,7 +14,7 @@ from pathlib import Path
 
 from selenium.common.exceptions import TimeoutException
 
-TIMEOUT = 60
+TIMEOUT = 120
 
 
 def start_dashboard() -> subprocess.Popen[str]:

--- a/tests/python/dashboard/utils.py
+++ b/tests/python/dashboard/utils.py
@@ -13,6 +13,8 @@ import time
 from pathlib import Path
 
 from selenium.common.exceptions import TimeoutException
+import os
+import time as _time
 
 TIMEOUT = 120
 
@@ -240,3 +242,25 @@ class DashboardTester:
             return null;
         """
         return self.sb.execute_script(js_script, state_name)
+
+
+def save_failure_screenshot(dashboard, request, directory: str | None = None) -> str | None:
+    """
+    Save a screenshot PNG for the current test and print the absolute path.
+
+    - Respects env var `IMPACTX_SCREENSHOT_DIR` to override the output folder.
+    - Falls back to `directory` arg or `screenshots/` in the current CWD.
+    Returns the absolute file path on success, else None.
+    """
+    base_dir = os.environ.get("IMPACTX_SCREENSHOT_DIR", directory or "screenshots")
+    try:
+        os.makedirs(base_dir, exist_ok=True)
+        ts = _time.strftime("%Y%m%d-%H%M%S")
+        # Keep name filesystem-friendly
+        name = f"{request.node.name}_{ts}.png".replace(os.sep, "_")
+        path = os.path.abspath(os.path.join(base_dir, name))
+        dashboard.sb.driver.save_screenshot(path)
+        print(f"Saved test screenshot: {path}")
+        return path
+    except Exception:
+        return None

--- a/tests/python/dashboard/utils.py
+++ b/tests/python/dashboard/utils.py
@@ -18,6 +18,7 @@ import pytest
 import time as _time
 
 TIMEOUT = 120
+APPROX_TOL = {"rel": 1e-12, "abs": 1e-12}
 
 
 def start_dashboard() -> subprocess.Popen[str]:
@@ -221,7 +222,7 @@ class DashboardTester:
 
                     v_num = None if value is None else float(value)
                     if v_num is not None and v_num == pytest.approx(
-                        float(expected_input), rel=1e-12, abs=1e-12
+                        float(expected_input), **APPROX_TOL
                     ):
                         return
                 except (TypeError, ValueError):

--- a/tests/python/dashboard/utils.py
+++ b/tests/python/dashboard/utils.py
@@ -122,6 +122,7 @@ class DashboardTester:
         """
         try:
             self.set_state("selected_lattice", element_name)
+            self.assert_state("is_selected_element_invalid", False)
             self.sb.click("#add_lattice_element")
         except Exception as error:
             raise Exception(

